src/gc/gc.d

```diff
@@ -2002,80 +2002,115 @@ struct Gcx
     Lagain:
         size_t pcache = 0;
 
+        // let dmd allocate a register for this.pools
+        auto pools = pooltable.pools;
+        const npools = pooltable.npools;
+        const minAddr = pooltable.minAddr;
+        const maxAddr = pooltable.maxAddr;
+
         //printf("marking range: [%p..%p] (%#zx)\n", p1, p2, cast(size_t)p2 - cast(size_t)p1);
-        for (; p1 < p2 && stackPos != stack.length; p1++)
+    Lnext: for (; p1 < p2; p1++)
         {
             auto p = cast(byte *)(*p1);
 
             //if (log) debug(PRINTF) printf("\tmark %p\n", p);
-            if (p >= pooltable.minAddr && p < pooltable.maxAddr)
+            if (p >= minAddr && p < maxAddr)
             {
                 if ((cast(size_t)p & ~cast(size_t)(PAGESIZE-1)) == pcache)
                     continue;
 
-                auto pool = findPool(p);
-                if (pool)
+                Pool* pool = void;
+                if (npools > 1)
                 {
-                    size_t offset = cast(size_t)(p - pool.baseAddr);
-                    size_t biti = void;
-                    size_t pn = offset / PAGESIZE;
-                    Bins   bin = cast(Bins)pool.pagetable[pn];
-                    void* base = void;
-
-                    //debug(PRINTF) printf("\t\tfound pool %p, base=%p, pn = %zd, bin = %d, biti = x%x\n", pool, pool.baseAddr, pn, bin, biti);
-
-                    // Adjust bit to be at start of allocated memory block
-                    if (bin < B_PAGE)
+                    size_t low = 0;
+                    size_t high = npools - 1;
+                    while (true)
                     {
-                        // We don't care abou setting pointsToBase correctly
-                        // because it's ignored for small object pools anyhow.
-                        auto offsetBase = offset & notbinsize[bin];
-                        biti = offsetBase >> pool.shiftBy;
-                        base = pool.baseAddr + offsetBase;
-                        //debug(PRINTF) printf("\t\tbiti = x%x\n", biti);
-
-                        if (!pool.mark.set(biti) && !pool.noscan.test(biti))
-                            stack[stackPos++] = Range(base, base + binsize[bin]);
+                        size_t mid = (low + high) >> 1;
+                        pool = pools[mid];
+                        if (p < pool.baseAddr)
+                            high = mid - 1;
+                        else if (p >= pool.topAddr)
+                            low = mid + 1;
+                        else break;
+
+                        if (low > high)
+                            continue Lnext;
```

Comment:
I would really prefer to keep it.

It might seem that it's not contributing, but it is helping. It's impact is proportional to the amount of times the outer branch (minAddr < p < maxAddr) branch is taken, right?

It's not hard to imagine legitimate cases where this branch would be taken very frequently:
-Very large pointer-based data structures, like trees, doubly linked lists, ...
-Unfortunate data-sets - i.e. 64-bit numbers that happen to be in range
-Both of the previous combined

One could still argue that when the outer branch is taken, the binary search is not the dominant factor, so I constructed a new test case to prove my theory: for every 16 bytes allocated, the first 8 bytes contain a pointer to a random address in range. The results show that this is indeed the case. The scan is now ~6x slower, because the outer branch is now taken much more frequently, and the inlined version is about ~25% faster than the non-inlined version.

Even in my original posted test case with no regard for memory content, the difference is ~1% which is still a pretty significant improvement, I think.